libs/patchwheel/patch.go

```diff
@@ -0,0 +1,280 @@
+func patchMetadata(r io.Reader, oldVersion, newVersion string) ([]byte, error) {
+	scanner := bufio.NewScanner(r)
+	var buf bytes.Buffer
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, versionKey) {
+			foundVersion := strings.TrimSpace(line[len(versionKey):])
+			if foundVersion != oldVersion {
+				return nil, fmt.Errorf("Unexpected version in METADATA: %s (expected %s)", strings.TrimSpace(line), oldVersion)
+			}
+			line = versionKey + newVersion
+		}
+		buf.WriteString(line + "\n")
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// patchRecord updates RECORD content: it replaces the old dist-info prefix with the new one
+// in all file paths and, for the METADATA entry, updates the hash and size.
+func patchRecord(r io.Reader, oldDistInfoPrefix, newDistInfoPrefix, metadataHash string, metadataSize int) ([]byte, error) {
+	scanner := bufio.NewScanner(r)
+	var newLines []string
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+		parts := strings.Split(line, ",")
+		if len(parts) < 3 {
+			// If the line doesn't have enough parts, preserve it as-is
+			newLines = append(newLines, line)
+			continue
+		}
+		origPath := parts[0]
+		if strings.HasPrefix(origPath, oldDistInfoPrefix) {
+			parts[0] = newDistInfoPrefix + origPath[len(oldDistInfoPrefix):]
+		}
+		// For the METADATA file entry, update hash and size.
+		if strings.HasSuffix(parts[0], "METADATA") {
+			parts[1] = "sha256=" + metadataHash
+			parts[2] = strconv.Itoa(metadataSize)
+		}
+		newLines = append(newLines, strings.Join(parts, ","))
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return []byte(strings.Join(newLines, "\n") + "\n"), nil
+}
+
+// PatchWheel patches a Python wheel file by updating its version in METADATA and RECORD.
+// It returns the path to the new wheel.
+// The version is updated according to the following rules:
+//   - if there is an existing part after + it is dropped
+//   - append +<mtime of the original wheel> to version
+//
+// The function is idempotent: repeated calls with the same input will produce the same output.
+// If the target wheel already exists, it returns the path to the existing wheel without redoing the patching.
+func PatchWheel(ctx context.Context, path, outputDir string) (string, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return "", err
+	}
+	wheelMtime := fileInfo.ModTime().UTC()
+
+	filename := filepath.Base(path)
+	wheelInfo, err := ParseWheelFilename(filename)
+	if err != nil {
+		return "", err
+	}
+
+	newVersion, newFilename := CalculateNewVersion(wheelInfo, wheelMtime)
+	outpath := filepath.Join(outputDir, newFilename)
+
+	if _, err := os.Stat(outpath); err == nil {
+		// Target wheel already exists, return its path
+		return outpath, nil
```

Comment:
Should we log it here that the wheel already there?

Reply:
added log message